### PR TITLE
SOL-151313: list-rdps aggregations 

### PR DESCRIPTION
**Pull request by helen-zhu-solace**
## Summary

Adds aggregation post-processing to `list-rdps` so the LLM can answer "which RDPs are down and why?" without iterating the raw list. Tool result now includes a `summary` block with `downCount`, `disabledCount`, `downButEnabledCount`, and `byLastFailureReason` alongside the existing raw RDP data.

Fixes SOL-151313

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Dependency update

## Motivation and Context

When a user asks an LLM "which RDPs are down?" against a VPN with many RDPs, the LLM previously had to fetch the full list and iterate it to compute counts and group by failure reason — wasting context tokens and risking arithmetic errors. This mirrors the same problem `list-queues` solved with its `noConsumerCount`/`congestedCount`/etc. summary block.

The aggregation also encodes two decisions the LLM shouldn't have to reinvent:
1. Distinguishing *unexpectedly down* (`downButEnabledCount`) from admin-disabled (`disabledCount`), which is the actual page-worthy signal.
2. `byLastFailureReason` restricts to RDPs that are down **and** enabled, so the map reflects unexpected active failures — not "RDP Shutdown" entries on intentionally-disabled RDPs, and not historical scars on RDPs that have since recovered.

## Changes Made

- New post-process handler `internal/composite/postprocess/handlers/list_rdps.go`, registered as `listRdps`. Mirrors the `listQueues` shape and robustness model (row-level skip on missing/wrong-type, structural error on shape mismatch).
- Aggregations: `downCount`, `disabledCount`, `downButEnabledCount`, `byLastFailureReason` (down && enabled && non-empty reason), plus the standard `scanned`, and `truncated`/`skipped` when applicable.
- New `boolField` helper alongside the existing `numField` / `stringField`.
- Unit tests `internal/composite/postprocess/handlers/list_rdps_test.go` — 8 test functions, including the load-bearing `TestListRdps_UnexpectedFailureFilter` (admin-disabled / recovered / empty-reason RDPs all stay out of the bucket), per-field skip, nil-field skip, truncation propagation, error paths, and empty-list shape.
- `internal/composite/definitions/tools.yaml`:
  - `list-rdps`: flipped `result.strategy` from `collect` to `postProcess: listRdps`; moved `select` from inline `args.select:` string to a top-level YAML list (required under `postProcess` strategy); appended a sentence to the description naming the summary fields and the active-and-enabled filter on `byLastFailureReason`.
  - `list-queues`: same documentation-only addition to its description, naming the four existing summary fields (drive-by parity fix so the two aggregating tools document themselves the same way).

## Testing

**Test Environment:**
- Go version: 1.25.11
- OS: Linux (Fedora 7.0.10-101.fc43)
- Broker version: `solace/solace-pubsub-standard:latest` (e2e-monitoring dockerized brokers)

**Tests Performed:**
- [x] Unit tests added/updated
- [ ] Integration tests added/updated (if applicable)
- [ ] E2E tests added/updated (covered by unit tests — see Test Coverage)
- [x] Tested locally with `go test -race ./...`
- [x] Tested with real Solace broker

**Test Coverage:**
- Mixed-state counts (healthy / down+enabled+reason / down+disabled+"RDP Shutdown" / disabled-only / duplicate reasons).
- Unexpected-failure filter: admin-disabled RDP with `"RDP Shutdown"` reason, recovered RDP with stale `lastFailureReason`, and down RDP with empty reason all stay out of `byLastFailureReason`.
- Empty list: all counts zero, `byLastFailureReason` an empty (not nil/absent) map, no `skipped`/`truncated` keys.
- Skip-row robustness: per-field missing (table-driven over each required field) and explicit-nil field both skip the row and tally `skipped` without dropping healthy rows' signals.
- Truncation propagation: `truncated: true` surfaces with `scanned`; omitted when false.
- Summary-key hygiene: `skipped` omitted when zero.
- Structural errors: missing step / data wrong type / item wrong type all hard-fail with clear messages.
- The validator cross-check (RequiredFields must appear in step `select`) is exercised generically by the existing `TestValidatePostProcess` and was confirmed in the live test (initial bad-YAML attempt failed boot with a precise error).
- **Manual live-broker run** against dockerized broker-a with two provisioned RDPs (`rdp-disabled` enabled=false, `rdp-down` enabled=true pointing nowhere): `summary` block returned the expected counts (`downCount:2`, `disabledCount:1`, `downButEnabledCount:1`) and after the design flip, `byLastFailureReason` correctly groups only the unexpectedly-down RDP's reason — the admin-disabled RDP's `"RDP Shutdown"` no longer appears in the bucket.
- E2E assertion intentionally skipped: the new logic is pure data transformation over a typed map (deterministically covered by unit tests); the YAML wiring is enforced at boot by `ValidatePostProcess`; the envelope shape is covered by `TestApplyResultStrategy_PostProcess_MergesSummary`. A live e2e assertion would re-prove what unit + boot-time checks already prove, and the "healthy RDP with stale failure reason" case can't be set up cleanly in a live broker without racing a connect/disconnect cycle.

## Breaking Changes

None. Tool envelope changes from `{rdps:{data,…}}` to `{rdps:{data,…}, summary:{…}}` — the existing `rdps` key and its contents are unchanged, so any caller reading `.rdps.data` keeps working.

**Impact:** N/A

**Migration Guide:** N/A

### Code Quality
- [x] Code follows the coding standards
- [x] Self-review completed
- [x] Code is well-commented (explains "why", not "what")
- [x] Complex logic has explanatory comments (unexpected-failure filter rationale, robustness vs. structural-error split)
- [x] No commented-out code or debug statements left in

### Security
- [x] **No credentials in code**
- [x] Followed secure logging rules (handler is pure data transformation — zero log calls; `/check-logs` clean)
- [x] Credential-carrying types implement `slog.LogValuer` (N/A — no new credential-carrying types)
- [x] No security vulnerabilities introduced (the two `golangci-lint` gosec hits exist on `main` in `internal/idpclient/client.go` and `internal/observability/correlation/middleware.go` — unrelated to this change)

### Testing
- [x] All tests pass locally for affected packages: `go test -race ./internal/composite/...`
- [x] No race conditions detected
- [x] Edge cases covered by tests
- [x] Error paths tested
- [x] Test names clearly describe what is being tested

### Documentation
- [ ] README.md updated (N/A — no user-facing surface change beyond the tool description, which is itself the documentation)
- [ ] docs/ updated (N/A)
- [x] godoc comments added/updated for exported functions
- [ ] CHANGELOG.md updated (N/A — repo does not maintain one)
- [ ] Examples updated (N/A)

### Git & CI
- [x] All commits are signed off (DCO: `git commit -s`)
- [x] Commits have clear, descriptive messages
- [x] Branch is up to date with main (rebased if needed)
- [x] No merge conflicts
- [x] CI checks passing (lint, build, test, E2E)

## Screenshots (if applicable)

```json
"summary": {
  "byLastFailureReason": { "No REST Consumers Up": 1 },
  "disabledCount":       0,
  "downButEnabledCount": 1,
  "downCount":           1,
  "scanned":             1
}

```
## Additional Notes

None.

## Related Issues/PRs

- Related to SOL-151343 (`get-rdp-status aggregations`) — companion ticket; mirrors the same active-and-enabled `byLastFailureReason` design decision.
- Pattern reference: previous PR that added `listQueues` post-process aggregations.

---

**For Reviewers:**

**Focus Areas:**
- `byLastFailureReason` semantics — the active-and-enabled filter is the load-bearing design choice. Confirm `TestListRdps_UnexpectedFailureFilter` covers what you'd want.
- YAML change: `select` moved from `args.select` string form to top-level YAML list — required under `postProcess` strategy and caught by the loader at boot.
- The description additions on `list-rdps` and `list-queues` are purely additive (existing wording untouched).

### CODE REVIEW

**Review by solace-aross — COMMENTED** [review]
Clean, well-scoped change. The handler mirrors the `listQueues` pattern faithfully and the test suite is genuinely thorough. `TestListRdps_UnexpectedFailureFilter` pins the load-bearing decision (admin-disabled, recovered, and empty-reason RDPs all stay out of the bucket) exactly where it belongs.

For whoever revisits the Copilot thread: all six comments target `byLastFailureReason` bucketing admin-disabled RDPs. The current code already filters on `!up && enabled && reason != ""`, so they're addressed. Safe to resolve.

One reconciliation before marking ready: the body still leads with "Waiting to test", but the testing section records a real-broker run with the expected counts (`downCount:2`, `disabledCount:1`, `downButEnabledCount:1`). Square those so the next reviewer trusts the checklist.

Solid PR.

File: internal/composite/postprocess/handlers/list_rdps.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	if t, _ := step["truncated"].(bool); t {
+		out["truncated"] = true
+	}
+	return out, nil
+}
+
+// boolField returns the bool at name and ok=true if present and bool-typed.
+// Missing, nil, or any other type returns ok=false so the caller can skip the
+// row rather than abort. Mirrors numField / stringField in list_queues.go.
+func boolField(item map[string]any, name string) (bool, bool) {
```
Comment (solace-aross) [inline internal/composite/postprocess/handlers/list_rdps.go]: Cross-PR collision to sort out before merge. #132 also adds a boolField with this exact signature to the same handlers package (list_vpns.go). boolField isn't on main today, so both PRs pass CI in isolation, but whichever merges second breaks the build with a redeclaration error.

Suggest hoisting numField/stringField/boolField into a shared fields.go in this package (in whichever of #128/#132 lands first) and dropping the local copy from the other. Four handlers share these coercion helpers now, so one home for them removes this whole class of conflict.

**Review by solace-aross — COMMENTED** [review]
Strong, well-tested change, and the byLastFailureReason filter is exactly right. Two things before merge, neither a defect in this PR on its own.

First, a build-level coordination issue with #132. See the inline note on boolField.

Second, a cross-tool convention question. Here downCount is !up, which includes admin-disabled RDPs, with downButEnabledCount as the clean signal. In #132 downCount is gated on enabled, so disabled VPNs land only in disabledCount. Copilot raised the same overlap on #132 and it was made exclusive there, so the two tools currently answer "what counts as down?" differently. An LLM reading across both gets "includes disabled" from one and "excludes disabled" from the other. I'd pick one rule for the set. My lean: gate downCount on enabled everywhere and keep downButEnabledCount as the explicit redundant signal, so "down" never silently folds in intentional shutdowns. Not blocking on its own, but worth settling before both merge.

**Comment by solace-aross** [comment]
Raised SOL-151500 to make the durable fix: consolidate numField/stringField/boolField into a shared fields.go so this collision class can't recur. You don't need to take that refactor on here.

For this PR, the only thing before merge is that just one of #128/#132 declares boolField. Whichever lands first keeps it; the other drops its copy and rebases. The permanent home is tracked in the ticket.

https://sol-jira.atlassian.net/browse/SOL-151500

**Review by solace-aross — COMMENTED** [review]
Merge-worthy on content. The design flip is in and pinned by TestListRdps_UnexpectedFailureFilter, and the body now reflects the live-broker run.

Holding formal approval until the rebase: #132 owns boolField now and lands first, so rebase here to drop the local copy before merging, otherwise whichever lands second fails on redeclaration. Ping me after the rebase and I'll approve.

One follow-up, not for this PR: downCount reads differently across the set, total not-up for RDPs vs enabled-and-down for VPNs. Arguably inherent to the RDP two-axis model, but an LLM reading both tools could conflate them. Worth settling at the naming/doc level.

**Review by solace-aross — APPROVED** [review]
Approving. The rebase resolves the boolField collision cleanly: the local declaration is dropped and list_rdps.go now uses the shared one from the merged list-vpns work. Single declaration in the package, CI green on the rebased head. The design-flip logic we discussed is intact (byLastFailureReason restricted to enabled-but-not-up).

The downCount naming divergence across list-rdps and list-vpns stays a follow-up, not a blocker for this.
